### ISSUE TITLE
azurerm_azuread_service_principal delay before being usable

### ISSUE THREAD

**Issue by mb290** [terraform-providers/terraform-provider-azurerm#1635]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

terraform -v
Terraform v0.11.7
+ provider.azurerm v1.10.0

### Affected Resource(s)

* terraform-provider-azurerm_v1.10.0_x4

### Terraform Configuration Files

```hcl
provider "azurerm" {
    version = "1.10.0"
 }

resource "azurerm_azuread_application" "test" {
  name                       = "exampleTFapplication"
  available_to_other_tenants = false
  oauth2_allow_implicit_flow = false
}

resource "azurerm_azuread_service_principal" "test" {
  application_id = "${azurerm_azuread_application.test.application_id}"
}

resource "azurerm_azuread_service_principal_password" "test" {
  service_principal_id = "${azurerm_azuread_service_principal.test.id}"
  value                = "BVcKK237/&&)hyz@%nsadasdsa(*&^CC#Nd3"
  end_date             = "2020-01-01T01:02:03Z"
}

resource "azurerm_resource_group" "test" {
  name     = "testResourceGroup1"
  location = "West US"
}

resource "azurerm_role_assignment" "test" {
    depends_on = ["azurerm_azuread_service_principal.test"]
  scope                = "${azurerm_resource_group.test.id}"
  role_definition_name = "Reader"
  principal_id         = "${azurerm_azuread_service_principal.test.id}"
}
```

### Panic Output

Error: Error applying plan:

1 error(s) occurred:

* azurerm_role_assignment.test: 1 error(s) occurred:

* azurerm_role_assignment.test: authorization.RoleAssignmentsClient#Create: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="PrincipalNotFound" Message="Principal ######################## does not exist in the directory #######-#####-######-#########."

(sensitive details have been hashed out).

### Expected Behavior

Here is a config that first creates the AzureAD application and the Service Principal. It then creates an RG followed by a role assignment. The logic here is we could have a single TF module that would allow us to on board new groups into an Azure subscription and generate them each their own SP. 

### Actual Behavior

When the azurerm_azuread_service_principal.test resource is created there looks to be a delay between creation and the ability to assign it it to a role and even with a depends_on that i've included in the sample code above that doesn't help. When I re-run the second time it always applies without issue as all other resources already exist.

### Steps to Reproduce

1. `terraform apply`

* #0000

**Comment by schoren** [comment]
I had the same issue today. In my case, I fixed it by using the `azurerm_azuread_application` id instead of the `azurerm_azuread_service_principal` id. Something like this:

```hcl
resource "azurerm_azuread_application" "test" {
  name                       = "exampleTFapplication"
  available_to_other_tenants = false
  oauth2_allow_implicit_flow = false
}

resource "azurerm_azuread_service_principal" "test" {
  application_id = "${azurerm_azuread_application.test.application_id}"
}

resource "azurerm_azuread_service_principal_password" "test" {
  service_principal_id = "${azurerm_azuread_service_principal.test.id}"
  value                = "BVcKK237/&&)hyz@%nsadasdsa(*&^CC#Nd3"
  end_date             = "2020-01-01T01:02:03Z"
}

resource "azurerm_resource_group" "test" {
  name     = "testResourceGroup1"
  location = "West US"
}

resource "azurerm_role_assignment" "test" {
  scope                = "${azurerm_resource_group.test.id}"
  role_definition_name = "Reader"
  principal_id         = "${azurerm_azuread_application.test.application_id}"
}
```

It's a weird behavior, but I got that from the `az ad sp create-for-rbac` command. When comparing to the Azure Portal, the actual ID used was the application ID.

Hope it helps!

**Comment by mb290** [comment]
@schoren  thanks for replying. I just tested this and when i tried the update I get the response:

Error: Error applying plan:

1 error(s) occurred:

* azurerm_role_assignment.test: 1 error(s) occurred:

* azurerm_role_assignment.test: authorization.RoleAssignmentsClient#Create: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="PrincipalNotFound" Message="Principal 408b56eeXXXXXXXXXXX does not exist in the directory #######-#####-######-#########."

I then confirmed the outputs and they are both different values:

Outputs:

azurerm_azuread_application_id = 408b56eeXXXXXXXXXXX 
azurerm_azuread_service_principal_id = b711cba7XXXXXXXXXXX 

I looked at the azurerm_role_assignment documentation and it does specifically call out the principal ID is required.

Am I missing something obvious?

**Comment by schoren** [comment]
Yes, yesterday I had a similar issue. I'm checking now to see if it is still happening. In another env, I had successfully deployed and assigned roles to services principal using that method.

**Comment by schoren** [comment]
Ok, now it's working with the original solution, using `azurerm_azuread_service_principal` id. Not sure why it worked different before, but it's working as expected now. Is it working for you?

**Comment by joakimhellum-in** [comment]
Hi, this is also the behavior of Azure CLI when using the command `az ad sp create-for-rbac`, as it pauses execution for 5 seconds and retries role assignment creation up to 36 times.
References:
https://github.com/Azure/azure-cli/blob/master/src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py#L941
https://github.com/Azure/azure-cli/blob/fc7d2ad961439ccd9b41547b0b6f9b8439cb99c5/src/azure-cli-core/azure/cli/core/commands/arm.py#L989
![image](https://user-images.githubusercontent.com/34519450/43163867-d303eb34-8f8f-11e8-8bf9-0b6923d50957.png)
Do anyone have any suggestion for workaround in terraform?

**Comment by joakimhellum-in** [comment]
We really want to avoid using the `local-exec` provisioner and sleep command as workaround, since we'd have to have pause execution approx. 180 seconds to really be sure server replication is done (sometimes server replication take long time). Also we run terraform on multiple OS/build agents where `sleep` is not always accessible. So it would be a really ugly hack. Using `az ad sp create-for-rbac` would be a better alternative for us than using the terraform resources currently.

**Comment by kjhosein** [comment]
@tombuildsstuff and/or anyone - would you clarify something for me?

It appears (to me at least) that the solution to the various `StatusCode=404, ErrorCode=ResourceNotFound` issues in the AzureRM provider is to code a fix/retry into the particular resource component. I've noticed multiple such issues here. 

Does this mean that you couldn't do something similar to the `max_retries` option in the AWS provider? 

Thanks for any insight!

**Comment by LaurentLesle** [comment]
I can confirm I have the same behaviour. This is related to the time to replicate the SP through the Azure AD servers. 

My scenario is:
- Create the azurerm_azuread_application,
- Create the azurerm_azuread_service_principal
- Create the azurerm_azuread_service_principal_password
- Create a Keyvault
- Assign a policy to that SP in KeyVault
- Connect to Azure RM provider using that SP to create a secret key

Get the error : AADSTS70001: Application with identifier 'app guid here' was not found in the directory

retry 1 min later another terraform apply and everything goes through.

**Comment by andresguisado** [comment]
I have tried with 30s, 60s,180s and 200s and I am still getting the same issue...

Using directly az-cli is what worked for me as  @joakimhellum-in mentioned previously:

```hcl
resource "azurerm_azuread_service_principal_password" "app_spn_password" {
  service_principal_id = "${azurerm_azuread_service_principal.app_spn_id.id}"
  value                = "${random_string.password.result}"
  end_date             = "${var.spn_end_date}" #2020-01-01T01:02:03Z  

  provisioner "local-exec" {
    command = "az role assignment create --role ${var.spn_role_definition_name} --assignee-object-id ${azurerm_azuread_service_principal.app_spn_id.id} --scope ${var.spn_scope}"
  }

}
```

**Comment by andresguisado** [comment]
Did anybody think to query the AD servers by PowerShell to see if the SPN has been replicated through and then carry on? 

http://community.idera.com/database-tools/powershell/ask_the_experts/f/active_directory__powershell_remoting-9/21621/check-if-user-exist-and-is-active-in-ad1-or-ad2

I am not sure if you can do this on Azure AD though...

**Comment by clstokes** [comment]
I'm getting `ServicePrincipalNotFound` errors for `azurerm_kubernetes_cluster` resources as well and a subsequent apply works. @tombuildsstuff, should I open a different issue than this one?

**Comment by tombuildsstuff** [comment]
@clstokes that sounds like the same underlying issue as this, so we can track that here. Thanks!

**Comment by logankp** [comment]
I'm getting the same issue but I'm not using depends_on. I created the cluster first then added the configuration to create the role assignment. No matter how many times I try to apply it fails.
